### DISCUSSION TITLE
Event System

### DISCUSSION THREAD

**Discussion by tac0turtle** [tendermint/tendermint#6114] [Q&A]
Recently the event system has been brought up more and more. This discussion is meant to outline some current issues and invite our community to provide feedback on how they use the event system and how they would like to see it improved. 

- Tendermint has its own custom query engine. The custom query engine has left many users hacking queries together in order to get what they need. There have been many issues opened through the years. https://github.com/tendermint/tendermint/issues?q=+label%3AC%3Aindexer+
- https://github.com/tendermint/tendermint/issues/5963
  - This issue is blocking the rosetta deployment. The rosetta API requires the ability to query supply changes in the sdk. These events are fired in Begin and/or endblock and are not queryable through the RPC. (https://github.com/tendermint/tendermint/issues/5133)
  - Secondly, when events are queried they arrive as a single large blob instead of individual events (https://github.com/tendermint/tendermint/issues/3665)

We (tendermint eng team) have had **brief** discussions on moving to a well-supported query engine, but before we make any decisions, we would like to hear from the users who are using the event system.

**Comment by alexanderbez** [comment]
/cc @ethanfrey I'm sure you have some input to add here.

Note, the approach I'd advocate for, which I shared on the call, is to consider entirely remove our proprietary events indexing and storage layer and simply rely on an embedded database with a rich JSON query language (e.g. sqlite) and expose the query layer via RPC and by directly proxying the query to the events database. We could collect events from the start to the end of the block, and create a record with `<BlockHeight, BlockHash, BlockEvents>`.

Question remains on how to (1) handle indexing and (2) what delivery and durability guarantees do we want to make?

**Reply by ethanfrey** [reply]
I will try to get time by tomorrow evening to write a detailed response, but a big :+1: for this:

> is to consider entirely remove our proprietary events indexing and storage layer and simply rely on an embedded database with a rich JSON query language

I was opposed to bike-shedding this from the start.

**Reply by alexanderbez** [reply]
Thanks @ethanfrey. Looking forward to your feedback and thoughts! I know you have tons of valuable first-hand experience with this aspect of Tendermint.

**Reply by tessr** [reply]
+1, thanks for your input, @ethanfrey, and I'm looking forward to your detailed response! I also have a strong preference for relying on a database (embedded or not!) that has solved this problem for us already.

**Reply by fdymylja** [reply]
Could you please elaborate and what is meant by delivery guarantees? Is it related only to pubsub acks? I imagine a successful tx (code=0) is gonna always have its events saved. So if I query a TX I should always get the events generated by it.

**Reply by alexanderbez** [reply]
No, delivery guarantees refer to reliability and availability of the data. i.e. All events from all blocks are guaranteed to be persisted, indexed, and queryable.

**Comment by jgimeno** [comment]
I think our needs for Rosetta are very simple, as long as we can get the events that happened on BeginBlock or EndBlock ordered as they happened I think is enough for us. So we leave the implementation details for you guys that really rock on this.

**Reply by tessr** [reply]
Do you have any requirements around e.g. delivery guarantees?

**Reply by fdymylja** [reply]
If by delivery guarantees we mean in pubsub, then we don't have this requirement, rosetta is not a stateful resource. It should be able to query a TX and get its event, same goes for endblock and beginblock events.

**Reply by alexanderbez** [reply]
No, that's not what that means. See https://github.com/tendermint/tendermint/discussions/6114#discussioncomment-377488

**Comment by ethanfrey** [comment]
Okay, I had a lot of thoughts on this, I will try to break it down into byte-size chunks and write them when I get the chance.

First thought - we need to separate the concerns of event recording and event indexing. I agree, we need a different/better indexer. But before we can index anything, let's look at the data that is produced and stored. Currently, the only data that is persisted outside of the transient pubsub system is the tx results. That is no Begin or EndBlock events are stored anywhere. And there is one event set per transaction. That is not ideal. But what do we want?

The event system has historically served two purposes - to (try to) watch for changes to state, and to look for the actions taken by the system. The first was "see if my balance has changed", which was approximated by listening to MsgSend or MsgWithdraw or MsgDelegate that are tied to my account. That can be done much better via the new [State Listening ADR 038](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-038-state-listening.md) and we should remove that concern from the event system. The other purpose is "inform me of what transition happened", and ideally tie it to the cause of that transition. eg. we get an event of a bank transfer from account A to account B, and that is tied to a transaction. We can look at the transaction and see if it was a swap, MsgSend, etc.

A third purpose which is a bit what "typed events' try to do, is to provide a way to index the transactions. "Find all stakng transactions". Do we consider that in scope? Some block explorers use events to search for tx, others index in their own db. 

Curious to opinions here as to what we are trying to capture? State, internal transitions, or transaction?

**Reply by alexanderbez** [reply]
This

> The other purpose is "inform me of what transition happened", and ideally tie it to the cause of that transition. eg. we get an event of a bank transfer from account A to account B, and that is tied to a transaction. We can look at the transaction and see if it was a swap, MsgSend, etc.

and 

> A third purpose which is a bit what "typed events' try to do, is to provide a way to index the transactions. "Find all staking transactions".

are what I believe the event system can and should accomplish, and, in fact, are not mutually exclusive in their concerns, design, or implementation.

**Reply by fdymylja** [reply]
> Curious to opinions here as to what we are trying to capture? State, internal transitions, or transaction?

Once [ADR-033](https://github.com/cosmos/cosmos-sdk/pull/7459) is implemented, in theory it could be possible to capture every state transition that happens (during delivertx, endblock, beginblock) as you could automatically use the message as event which is causing the state change (the message passes through the grpc router which could automatically generate an event containing the field of the message processed).

This becomes really important in endblock and beginblock as it currently is impossible to understand what state changes happen during those phases as they're not represented as transactions.

**Comment by ethanfrey** [comment]
Sorry for the long silence. Adding a few thoughts on use cases (each can be it's own thread).

Use Case #1: Responsive Clients:

When building a wallet you would like the balance to update when the user receives a new incoming Tx. You would like the governance UI to add new votes to the list as they are made, etc.  This is not just for Wallets, however, imagine if you made a client-side "liquid democracy" app (perhaps using subkeys) that listened for votes on a proposal and when some threshold of people you follow vote, it will submit a tx for your vote.

In either case, you can either continually poll the state (which will kill an RPC server if 1000s of clients are doing this), or subscribe to events. We should ensure that event subscriptions work well and that they can detect such changes. They don't need all info in the event, just a signal, and just query state again when the event arises.

**Reply by ethanfrey** [reply]
This is probably the simplest of the 3 use cases, but we do need to make some thoughts on reliable/reconnectable event subscriptions. We did quite some work on this with iov-core and weave, and often had the case that websocket connections would be dropped after say 15 minutes in a background tab, and then when you come back to the app, it no longer responded to changes properly. We then built some wrapper to "QueryAndSubscribe" and make the same query for `height > H` (where we know we had seen H before) as well as making a subscription, then combining them into one stream.

This was a bit tricky and especially handling all the layers and handling websocket reconnects properly. While re-envisioning the event system, maybe we can look at making this easier to handle (either with websockets or gRPC, they both will have reconnect issues). If you look at "guarantee once delivery" eg. PostgreSQL replogs or most messaging queue, a listener/subscriber gets a slot, and then marks the message as "seen" when it has been seen. Unseen messages wait in the queue (up to some time/quantity limit) and on a reconnect, the client easily carries on from the last seen item.

This may be out of scope of any current patches, but should at least be considered if you will do a redesign of the event system.  Especially if you think to use a DB or messaging service to make this easier to build

**Reply by alexanderbez** [reply]
I put this under the durability and availability bucket of the new event system, more or less separate from the underlying indexing and persistence mechanism, but yes, indeed, a very valid point to raise.

**Comment by ethanfrey** [comment]
Use Case #2: Block Explorers

We want to make transactions visible to a user. Primarily, all transactions they signed, what they included and whether or not they succeeded. It will obviously need to be able to parse the transaction and all messages, either with chain-dependent logic to parse all protobuf messages, or as suggested above, by the event system outputting a JSON version of the messages. 

However, they often also want to show all "incoming" transactions. This is obviously subjectively defined, but generally includes MsgSend that affects a user. A validator may want to see a list of all Delegate/Undelegate transactions from others that affect them. And when their is an on-chain orderbook, fulfillment of those orders (which often will not happen in the same tx that I submitted) should somehow show up, as an "incoming tx" easy to find from my tx that made the order (even when the match was made in EndBlock)

We also need to make sure that we can separate by *message*, not just by *Tx*. There is a lot of work going on now for more and more sophisticated clients and we should start expecting multi-msg tx (even with one signer) to become normal. Atomicity and ordering can be a huge boon in DeFi.

**Reply by alexanderbez** [reply]
This falls more under the application's responsibility, no? Specifically, how the application decides to construct the composite/typed event, e.g. `message.sender` or `message.recipient`.

**Comment by ethanfrey** [comment]
Use Case #3: Auditing  

When filing taxes, you need to declare the source of all income. You may show the list of incoming and outgoing MsgSend and the genesis balance and somehow they don't add up to the actual balance. Ah... staking rewards. DeFi returns, etc.  We need a way to review all messages that affected your balance in any way.

There is currently a `"type": "transfer"` Event triggered on everything that calls `bank/Keeper.Send` This is good and can be used to find more of this data. However, we need to ensure this captures all events (incl. Begin/EndBlock) and also can be tied to the original cause. For this, we will need some sort of tx/msg list for Begin/EndBlock, or at least some kind of grouping. Maybe just when the module.EndBlocker is called, it makes a new stream of events. So rather than grouping them into list of events by message (which is done properly in the sdk before being mangled in Tendermint), we could have a list of events grouped by module. The first event could just say which module (eg. "distribution" or "gravity-dex") and then all actions there would have their sub-events. 

For eg. distribution, finding a  `transfer` event going to your account in EndBlocker labeled 'distribution' should be enough to classify this as a staking reward. For other modules, such as the dex, you need to link this to not just to the dex module, but which order was closed (as mentioned above), so some more structure is needed... maybe each module emits not just a list of Events, but rather a list of Messages, each with their own description and list of events?

**Reply by alexanderbez** [reply]
What if when you query for a block or subscribe to a new block event, you get the block with all events in begin block, end block and all deliver tx events?

**Reply by fdymylja** [reply]
> When filing taxes, you need to declare the source of all income. You may show the list of incoming and outgoing MsgSend and the genesis balance and somehow they don't add up to the actual balance. Ah... staking rewards. DeFi returns, etc. We need a way to review all messages that affected your balance in any way.

Currently with https://github.com/cosmos/cosmos-sdk/pull/8473 and https://github.com/cosmos/cosmos-sdk/pull/8656

Balance can now be tracked by following coin_spend and coin_received events, they cover all the cases in which balance changes, at begin and endblock level too.

>  maybe each module emits not just a list of Events, but rather a list of Messages, each with their own description and list of events?

If intermodule communication ADR is implemented, and the keepers abandoned when it comes to interaction between modules, it could be possible to use the module-to-module router to emit events based on the msg it is routing: example MsgStaking causes (internally) a MsgBankSend etc.

**Reply by alexanderbez** [reply]
> If intermodule communication ADR is implemented, and the keepers abandoned when it comes to interaction between modules, it could be possible to use the module-to-module router to emit events based on the msg it is routing: example MsgStaking causes (internally) a MsgBankSend etc.

Sure, but that seems orthogonal to this discussion, no?

**Reply by fdymylja** [reply]
> Sure, but that seems orthogonal to this discussion, no?

Maybe, it was just to add more context on the possible use-cases that might come up from the sdk (which is tendermint's main consumer). Not saying that the work on events should revolve around the sdk, but still..

**Reply by ethanfrey** [reply]
> What if when you query for a block or subscribe to a new block event, you get the block with all events in begin block, end block and all deliver tx events?

Good question. Ideally there are multiple levels of granularity (something equivalent to tx) and you only get the chunk that matches. Maybe each Module.BeginBlocker is counted as one "Tx" and then you only get back the modules that matched. And each modules' events may be broken down in a list of "messages", eg. [][]Event (two levels of nesting - first group is an intention, eg. close order X or pay staking to Y, under that the list of all events that came from the given intention).

Clearly this will need some changes in the sdk and likely the apps, but it requires a different API to use. For compatibility, the sdk could just split by module and we stick all events from the module under "msg[0]". If the module wants to, it could split into multiple "messages".

**Comment by ethanfrey** [comment]
To properly capture all the above use-cases in a "Event system", we will need to output even more data than now, and output it in a more structured format. Currently, in CosmJS, while we make queries in the event system, we do not use those events to reconstruct any state (eg. for a block explorer). The `raw_log` actually contains the events broken down by message in a nice JSON format, and not collapsed together (eg. 3 transfer messages in one sdk.Msg as listed separately).

This makes me feel that we could (1) drop the abci.Event type completely and *not* try to deal with this in Tendermint in some generic format that could never understand multiple messages per Tx and (2) use a Cosmos-SDK aware indexer to process this JSON output.

This Cosmos-SDK indexer either be a module optionally enabled (eg only on rpc nodes, not on validators) or a third-party process that listens to some stream, as proposed in [ADR 033](https://github.com/cosmos/cosmos-sdk/pull/7459). I would favor an optional out-of-process indexer, and provide a simple implementation that ships with the SDK along with a very clear and simple API. And allow for more advanced ones (eg. with GraphQL support), that ideally *extend* the API to provide more functionality, but are compatible with clients for the simple one.

I would actually suggest to build one with eg. PostgreSQL and a messaging queue (RabbitMQ, Redis, etc), but maybe that is for the advanced class

**Comment by ethanfrey** [comment]
Performance:

We currently write quite a few bytes both in the `raw_log` field as well as the event history. This is not charged gas at all, yet is stored in Tendermint along with the transactions themselves (which cost 2 gas/byte).

Because of that you can see PRs like https://github.com/cosmos/cosmos-sdk/pull/8624 which will dump the header and entire precommit set (125 sigs?) in JSON into the event system, as it (a) is "free" and (b) doesn't break consensus. I feel this is an abuse of the event system, and we have flagged this some months ago as a potential vulnerability in CosmWasm https://github.com/CosmWasm/wasmd/issues/297

I see two ways to handle this:

(1) Charge a gas fee per byte in raw_log and (normalized) Events. This would limit some "excessive" usage of the event system, but also have the effect that any change to emitted events is Consensus breaking (as the message would consume different gas on two different nodes)

(2) Make a new Tendermint option to avoid writing the log/event data. This could be enabled on validators to minimize disk usage, but disabled on any query node (that wishes to provide that info to clients). This could cause issues with syncing nodes however, as if you sync from a node with the "ignore_logs=true", then your node will never get the logs from historical synced data (only from the blocks you execute) if you have "ignore_logs=false" 

Even with option (2) there should be some limit to prevent a tx from writing eg. 4MB of event data. We will add our own controls in CosmWasm, but in general, the SDK should provide some checks here.

**Reply by alexanderbez** [reply]
Totally agree that the events system can be abused. I think there might be some discussions to have around charging gas for events. But note, that for (2) I believe you can already do this, see [here](https://github.com/cosmos/cosmos-sdk/blob/master/server/config/config.go#L68-L70). The application can tell Tendermint what events it wants to index and persist.
